contracts/modules/staking/PooledStaking.sol

```diff
@@ -489,28 +500,104 @@ contract PooledStaking is MasterAware, IPooledStaking {
     address contractAddress, uint amount
   ) public onlyInternal whenNotPausedAndInitialized noPendingBurns {
 
+    address[] memory contractAddresses = new address[](1);
+    contractAddresses[0] = contractAddress;
+    _pushRewards(contractAddresses, true);
+
     burn.amount = amount;
     burn.burnedAt = now;
     burn.contractAddress = contractAddress;
 
     emit BurnRequested(contractAddress, amount);
   }
 
-  function pushReward(address contractAddress, uint amount) external onlyInternal whenNotPausedAndInitialized {
+  function _getCurrentRewardsRound() internal view returns (uint) {
+
+    uint roundDuration = REWARD_ROUND_DURATION;
+    uint startTime = REWARD_ROUNDS_START;
+
+    require(startTime != 0, 'REWARD_ROUNDS_START is not initialized');
+
+    return now <= startTime ? 0 : (now - startTime) / roundDuration;
+  }
+
+  function getCurrentRewardsRound() external view returns (uint) {
+    return _getCurrentRewardsRound();
+  }
+
+  /**
+   * @dev Pushes accumulated rewards to the processing queue.
+   */
+  function _pushRewards(address[] memory contractAddresses, bool skipRoundCheck) internal {
+
+    uint currentRound = _getCurrentRewardsRound();
+    uint lastRewardIdCounter = lastRewardId;
+    uint pushedRewards = 0;
+
+    for (uint i = 0; i < contractAddresses.length; i++) {
+
+      address contractAddress = contractAddresses[i];
+      ContractReward storage contractRewards = accumulatedRewards[contractAddress];
+      uint amount = contractRewards.amount;
+
+      bool canPush = amount > 0 && (skipRoundCheck || currentRound > contractRewards.lastDistributionRound);
+
+      if (!canPush) {
+        continue;
+      }
+
+      rewards[++lastRewardIdCounter] = Reward(amount, now, contractAddress);
+      emit RewardRequested(contractAddress, amount);
+
+      contractRewards.amount = 0;
+      contractRewards.lastDistributionRound = currentRound;
+      ++pushedRewards;
+    }
+
+    if (pushedRewards != 0) {
 
-    rewards[++lastRewardId] = Reward(amount, now, contractAddress);
+      lastRewardId = lastRewardIdCounter;
 
-    if (firstReward == 0) {
-      firstReward = lastRewardId;
+      if (firstReward == 0) {
+        firstReward = lastRewardIdCounter;
```

Comment:
there could be many rewards pushed as part of this loop, shouldn't this be the first value that was used as a reward id in that loop, instead of the last?